Feat: Add final states

deuslirio opened the pull request:
Add a feature to explicitly define which states are final states.

This requires avoiding transitions coming from final states and also makes it possible to list all final states of an FSM.

The Java Spring-statemachine is an example of FSM implementation which allow final states definition:
https://docs.spring.io/spring-statemachine/docs/current/reference/#statemachine-config-states
```
...
	.withStates()
		.initial(States.S1)
		.end(States.SF) // Final State definition
		.states(EnumSet.allOf(States.class));
...
```

fgmacedo commented:
@deuslirio Nice! Thanks for your time implementing this.

Do you think we should add validations on `StateMachine.check` to ensure proper machine definitions?

Like:

1. Final state does not should have defined transitions starting from that state>
2. What else?

deuslirio commented:
Thanx 👍️
I have removed the validation from `can_run` and added it in `check`. 
It will raise exception in FSM creation instead of raise in transition execution.

fgmacedo commented on statemachine/statemachine.py:
```diff
@@ -364,12 +369,24 @@ def check(self):
+            state for state in self.final_states if state.transitions
+        ]
+
+        if final_state_with_invalid_transitions:
+            raise InvalidDefinition(_('Final state does not should have defined '
+                                      'transitions starting from that state'))
```
Nice!
